include/ua_plugin_securitypolicy.h

```diff
@@ -0,0 +1,424 @@
+     * \param channelContext the channelContext which contains information about the keys to encrypt data.
+     * \param data the data that is encrypted. The encrypted data will overwrite the data that was supplied.
+     */
+    UA_StatusCode (*const encrypt)(const UA_SecurityPolicy *securityPolicy,
+                                   const void *endpointContext,
+                                   const void *channelContext,
+                                   const UA_ByteString *data);
+    /**
+     * \brief Decrypts the given cyphertext using an asymmetric algorithm and key.
+     *
+     * \param securityContext the SecurityContext which contains information about the keys needed to decrypt the message.
+     * \param data the data to decrypt. The decryption is done in place.
+     */
+    UA_StatusCode (*const decrypt)(const UA_SecurityPolicy *securityPolicy,
+                                   const void *endpointContext,
+                                   UA_ByteString *data);
+    /**
+     * Generates a thumprint for the specified certificate using a SHA1 digest
+     *
+     * \param certificate the certificate to make a thumbprint of.
+     * \param thumbprint an output buffer for the resulting thumbprint. Always
+                         has the length specified in the thumprintLenght in the asymmetricModule.
+     */
+    UA_StatusCode (*const makeThumbprint)(const UA_SecurityPolicy *securityPolicy,
+                                          const UA_ByteString *certificate,
+                                          UA_ByteString *thumbprint);
+
+    /**
+     * \brief Calculates the padding size for a message with the specified amount of bytes.
+     *
+     * \param securityPolicy the securityPolicy the function is invoked on.
+     * \param channelContext the channel context that is used to obtain the plaintext block size.
+     *                       Has to have a remote certificate set.
+     * \param bytesToWrite the size of the payload plus the sequence header, since both need to be encoded
+     * \param paddingSize out parameter. Will contain the paddingSize byte.
+     * \param extraPaddingSize out parameter. Will contain the extraPaddingSize. If no extra padding is needed, this is 0.
+     * \return the total padding size consiting of high and low byte.
+     */
+    UA_UInt16 (*const calculatePadding)(const UA_SecurityPolicy *securityPolicy,
+                                        const void *channelContext,
+                                        const void *endpointContext,
+                                        size_t bytesToWrite,
+                                        UA_Byte *paddingSize,
+                                        UA_Byte *extraPaddingSize);
+
+    const size_t minAsymmetricKeyLength;
+    const size_t maxAsymmetricKeyLength;
+    const size_t thumbprintLength;
+
+    const UA_SecurityPolicySigningModule signingModule;

```

Comment:
Are you sure that the signingModule can be different for symmetric / asymmetric encryption?
Otherwise put in a top-level module.